java-frontend/src/main/java/org/sonar/java/ast/visitors/FileLinesVisitor.java

```diff
@@ -42,40 +49,98 @@
 public class FileLinesVisitor extends SubscriptionVisitor {
 
   private final SonarComponents sonarComponents;
-  private final Charset charset;
-  private final Set<Integer> linesOfCode = Sets.newHashSet();
-  private final Set<Integer> linesOfComments = Sets.newHashSet();
+  private final Set<Integer> linesOfCode = new HashSet<>();
+  private final Set<Integer> linesOfComments = new HashSet<>();
+  private final Set<Integer> executableLines = new HashSet<>();
 
-  public FileLinesVisitor(SonarComponents sonarComponents, Charset charset) {
+  public FileLinesVisitor(SonarComponents sonarComponents) {
     this.sonarComponents = sonarComponents;
-    this.charset = charset;
   }
 
   @Override
   public List<Tree.Kind> nodesToVisit() {
+    if(sonarComponents.isSQGreaterThan62()) {
+      return ImmutableList.of(Tree.Kind.TOKEN, Tree.Kind.METHOD, Tree.Kind.CONSTRUCTOR, Tree.Kind.INITIALIZER, Tree.Kind.STATIC_INITIALIZER, Tree.Kind.VARIABLE);
+    }
     return ImmutableList.of(Tree.Kind.TOKEN);
   }
 
   @Override
   public void scanFile(JavaFileScannerContext context) {
     super.scanFile(context);
-
-    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(context.getFile());
-    // TODO minimize access to files, another one in LinesVisitor
-    int fileLength;
-    try {
-      fileLength = Files.readLines(context.getFile(), charset).size();
-    } catch (IOException e) {
-      throw Throwables.propagate(e);
-    }
+    File currentFile = context.getFile();
+    FileLinesContext fileLinesContext = sonarComponents.fileLinesContextFor(currentFile);
+    int fileLength = sonarComponents.fileLength(currentFile);
     for (int line = 1; line <= fileLength; line++) {
       fileLinesContext.setIntValue(CoreMetrics.NCLOC_DATA_KEY, line, linesOfCode.contains(line) ? 1 : 0);
       fileLinesContext.setIntValue(CoreMetrics.COMMENT_LINES_DATA_KEY, line, linesOfComments.contains(line) ? 1 : 0);
+      if(sonarComponents.isSQGreaterThan62()) {
+        fileLinesContext.setIntValue(CoreMetrics.EXECUTABLE_LINES_DATA_KEY, line, executableLines.contains(line) ? 1 : 0);
+      }
     }
     fileLinesContext.save();
 
     linesOfCode.clear();
     linesOfComments.clear();
+    executableLines.clear();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    List<? extends Tree> trees = null;
+    switch (tree.kind()) {
+      case INITIALIZER:
+      case STATIC_INITIALIZER:
+        trees = ((BlockTree) tree).body();
+        break;
+      case VARIABLE:
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this switch case number of lines from 13 to at most 5, for example by extracting code into methods. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1151)